File: prac1_sorts/c++ src/heapSort.cpp
#include <iostream>
#include "Timer.hpp"

using namespace std;
void swap(int *arr,int a,int b){
    int tmp=arr[a];
    arr[a]=arr[b];
    arr[b]=tmp;
}
void heapify(int *arr,int n, int i){
    int largo=i;
    int l=2*i+1;
    int r=2*i+2;
    if(l<n && arr[l]>arr[largo]){
        largo=l;
    }
    if(r<n && arr[r]>arr[largo]){
        largo=r;
    }
    if(largo!=i){
        swap(arr,i,largo);
        heapify(arr,n,largo);
    }
}
void heapSort(int *arr,int n){
    for(int i=(n/2)+1;i>=0;i--){
        heapify(arr,n,i);
    }
    for(int i=n-1;i>0;i--){
        swap(arr,0,i);
        heapify(arr,i,0);
    }
}
void print(int *arr,int n){
    for(int i=0;i<n;i++){
        cout<<arr[i]<<" ";
    }
    cout<<endl;
}

int main()
{
    int numLengths;
    cin>>numLengths;

    for (int i = 0; i < numLengths; ++i)
    {
        float avg = 0;
        int numArrays = 10;
        int n;
        cin>>n;
        for(int j = 0; j < numArrays; ++j)
        {
            int* arr = new int [n]{0};
            for (int k = 0; k < n; ++k)
                cin>>arr[k];
            Timer timer;
            timer.start();
            heapSort(arr,n);
            timer.stop();
            avg += timer.elapsedMilliseconds();
            delete [] arr;
        }
        avg /= numArrays;
        cout << n << " " << avg << endl;
    }
    

    return 0;
}

File: prac1_sorts/python src/heapsort.py

def heapify(arr, n, i):
    largo = i
    l = (2*i)+1
    r = (2*i)+2
    if l < n and arr[l] > arr[largo]:
        largo = l
    if r < n and arr[r] > arr[largo]:
        largo = r
    if largo != i:
        arr[i], arr[largo] = arr[largo], arr[i]
        heapify(arr, n, largo)


def heapsort(arr, n):
    for i in range(n//2+1, -1, -1):
        heapify(arr, n, i)
    for j in range(n-1, 0, -1):
        arr[0], arr[j] = arr[j], arr[0]
        heapify(arr, j, 0)


array = [64, 34, 25, 12, 22, 11, 90, 23]
na = len(array)
heapsort(array, na)
for i in range(na):
    print("%d" % array[i]),


